File: kotlin/advent/src/main/kotlin/org/gristle/pdxpuzzles/advent/y2019/Y19D18.kt
package org.gristle.pdxpuzzles.advent.y2019

import org.gristle.pdxpuzzles.advent.utilities.Day
import org.gristle.pdxpuzzles.utilities.algorithms.Graph
import org.gristle.pdxpuzzles.utilities.algorithms.Graph.steps
import org.gristle.pdxpuzzles.utilities.objects.Grid
import org.gristle.pdxpuzzles.utilities.objects.toGrid
import org.gristle.pdxpuzzles.utilities.objects.toMutableGrid

class Y19D18(input: String) : Day {

    /**
     * Used for the edgemap, contains enhanced edge data. This includes the location of the edge, but also all
     * keys needed to be collected to make the move (precedingKeys) as well as all keys that are encountered while
     * traveling to the key (interveningKeys). precedingKeys is used to filter out moves that are not yet possible
     * due to not having the right key to pass through a door. interveningKeys is used to update the state so that
     * all keys are added in one swoop.
     */
    data class Key(val currentKey: Char, val precedingKeys: Set<Char>, val interveningKeys: Set<Char>)

    /**
     * Used for the main Dijkstra search. Contains the current position of each robot (location) and all keys that
     * have been collected up to that point (keys).
     */
    data class State(val location: List<Char>, val keys: Set<Char>)

    private val tunnels = input.toGrid()

    fun solve(tunnels: Grid<Char>, starts: String): Int {

        // Intermediate edge map used to later make the keyEdges edge map used for the final solution. Just gives
        // the distance to the nearest keys and doors. Once a key or door is reached, the BFS does not explore 
        // beyond that key or door. But it will still look for other keys or doors that do not require going through
        // that key or door.
        val basicEdges: Map<Char, List<Graph.Edge<Char>>> = buildMap {
            tunnels // take input...
                .withIndex() // associate each value with its index...
                .filter { (_, value) -> value.isLetter() || value in starts } // only run for important positions
                .forEach { (index, start) ->

                    // lambda for getting edges from the grid
                    val getEdges: (IndexedValue<Char>) -> List<IndexedValue<Char>> = { (index, current) ->

                        // If a key or door is reached, end this particular path.
                        if (current.isLetter() && current != start) {
                            emptyList()
                        } else { // ...otherwise keep exploring
                            tunnels
                                .getNeighborsIndexedValue(index) // get adjacent values
                                .filter { it.value != '#' } // remove walls
                        }
                    }

                    // populate edge map with all key and door locations that do not require passing through a 
                    // different key or door to get to
                    this[start] = Graph.bfs(IndexedValue(index, start), defaultEdges = getEdges)
                        .drop(1)
                        .filter { it.id.value.isLetter() } // only look at keys or doors
                        .map { Graph.Edge(it.id.value, it.weight) } // convert to an Edge
                }
        }

        // Edge map used for the main Dijkstra search. Itself uses a Dijkstra search relying on the basic edge map.
        val keyEdges: Map<Char, List<Graph.Edge<Key>>> = basicEdges
            .keys
            .filter { it.isLowerCase() || it in starts }
            .associateWith { key -> // make edge map
                Graph
                    .dijkstraSequence(key, edges = basicEdges)
                    .drop(1) // first edge returned is the start, so drop that 
                    .filter { it.id.isLowerCase() } // only look at key locations
                    .map { distance ->

                        // get the full path to key, dropping the start
                        val path = distance.path().drop(1)

                        // separate doors from keys
                        val (keyList, interveningKeyList) = path.partition { it.id.isUpperCase() }
                        val keys = keyList
                            .map { it.id.lowercaseChar() }
                            .toSet()
                        val interveningKeys = interveningKeyList
                            .map { it.id }
                            .toSet()

                        // create edge with metadata
                        Graph.Edge(Key(path.last().id, keys, interveningKeys), path.last().weight)
                    }.toList()
            }

        // initial state for the main Dijkstra algorithm
        val startState = State(starts.toList(), emptySet())

        // continue Dijkstra until all keys have been collected
        val endCondition = { state: State -> state.keys.size == 26 }

        // lambda finds edges by looking at each robot position, looking where each could move, and generating
        // new states for each possible move
        val findEdges = { state: State ->
            state.location.flatMapIndexed { robot, location ->
                keyEdges
                    .getValue(location) // gets all edges for location
                    .filter { keyEdge -> // keeps if key is new to state and all prerequisite keys collected 
                        keyEdge.vertexId.currentKey !in state.keys
                                && state.keys.containsAll(keyEdge.vertexId.precedingKeys)
                    }.map { edge -> // converts valid edge to state

                        // updates robot locations
                        val newLocations: List<Char> =
                            state.location.toMutableList().apply { this[robot] = edge.vertexId.currentKey }

                        // updates keys collected
                        val newKeys = state.keys union edge.vertexId.interveningKeys
                        val newState = State(newLocations, newKeys)
                        Graph.Edge(newState, edge.weight)
                    }
            }
        }

        return Graph.dijkstra(startState, endCondition, defaultEdges = findEdges).steps()
    }

    override fun part1() = solve(tunnels, "@")

    override fun part2(): Int {
        val robots = "@$%^"

        // Change the grid so that there are 4 quadrants
        val quadrants: Grid<Char> = tunnels.toMutableGrid().apply {
            val originalStart = indexOf('@')
            val wallIndices = getNeighborIndices(originalStart)
            val newStartIndices = getNeighborIndices(originalStart, true) - wallIndices.toSet()
            this[originalStart] = '#'
            wallIndices.forEach { this[it] = '#' }
            newStartIndices.forEachIndexed { index, i -> this[i] = robots[index] }
        }
        return solve(quadrants, robots)
    }
}

fun main() = Day.runDay(Y19D18::class)

//    Class creation: 4ms
//    Part 1: 3918 (123ms)
//    Part 2: 2004 (141ms)
//    Total time: 270ms

File: rust/advent/src/bin/2019/y19d18.rs
use advent::utilities::get_input::get_input;
use rustc_hash::{FxBuildHasher, FxHashMap};
use std::collections::HashMap;
use std::fmt::{Debug, Formatter};
use itertools::Itertools;
use utilities::graphs::{bfs, dijkstra, no_end_condition, EdgeInfo, PathInfo};
use utilities::structs::grid::Grid2;
use utilities::structs::stopwatch::{ReportDuration, Stopwatch};

type Input = Grid2<char>;
type Output = usize;
type CharSet = [bool; 26];

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let input = get_input(19, 18).unwrap();
    let input = parse_input(&input);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

/**
 * Used for the edgemap, contains enhanced edge data. This includes the location of the edge, but
 * also all keys needed to be collected to make the move (precedingKeys) as well as all keys that
 * are encountered while traveling to the key (interveningKeys). precedingKeys is used to filter out
 * moves that are not yet possible due to not having the right key to pass through a door.
 * interveningKeys is used to update the state so that all keys are added in one swoop.
 */
#[derive(Clone, Copy)]
struct Key {
    current: char,
    preceding: CharSet,
    intervening: CharSet,
}

impl Debug for Key {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let p = self.preceding.iter().enumerate()
            .filter(|&(_, &b)| b)
            .map(|(i, _)| (i + 'a' as usize) as u8 as char)
            .join(",");
        let i = self.intervening.iter().enumerate()
            .filter(|&(_, &b)| b)
            .map(|(i, _)| (i + 'a' as usize) as u8 as char)
            .join(",");
        format!("{:?}, pre: {p}, inter: {i}", self.current).fmt(f)
    }
}

/**
 * Used for the main Dijkstra search. Contains the current position of each robot (location) and all
 * keys that have been collected up to that point (keys).
 */
#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
struct State {
    location: Vec<char>,
    keys: CharSet,
}

impl Debug for State {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let k = self.keys.iter().enumerate()
            .filter(|&(_, &b)| b)
            .map(|(i, _)| i + 'a' as usize)
            .join(",");
        format!("{:?}: {k}", self.location).fmt(f)
    }
}

fn parse_input(input: &str) -> Input {
    input.try_into().unwrap()
}

fn solve(tunnels: &Input, robots: Vec<char>) -> Output {
    // Intermediate edge map used to later make the keyEdges edge map used for the final solution.
    // Just gives the distance to the nearest keys and doors. Once a key or door is reached, the BFS
    // does not explore beyond that key or door. But it will still look for other keys or doors that
    // do not require going through that key or door.
    let mut basic_edges: HashMap<char, Vec<(char, usize)>, FxBuildHasher> = FxHashMap::default();
    for (index, &start) in tunnels.iter()
        .enumerate()
        .filter(|&(_, c)| c.is_ascii_alphabetic() || robots.contains(c))
    {
        let get_edges = |_: EdgeInfo<usize>, (index, current): &(usize, char)|
            -> Vec<(usize, char)>
        {
            // If a key or door is reached, end this particular path.
            if current.is_ascii_alphabetic() && *current != start {
                Vec::new()
            } else {
                tunnels.adjacent(*index, false)
                    .unwrap()
                    .filter(|adj| *adj.value != '#')
                    .map(|adj| (adj.index, *adj.value))
                    .collect()
            }
        };

        let edges = bfs((index, start), get_edges, no_end_condition);

        let edges = edges
            .nodes
            .into_iter()
            .skip(1)
            .filter(|&((_, c), _)| c.is_ascii_alphabetic())
            .map(|((_, c), info)| {
                (c, info.cost)
            })
            .collect();
        basic_edges.insert(start, edges);
    }

    let key_edges: FxHashMap<char, Vec<(Key, usize)>> = basic_edges
        .keys()
        .filter(|&&key| key.is_lowercase() || robots.contains(&key))
        .map(|&key| {
            let edges: PathInfo<char, usize> = dijkstra(
                key,
                |_, c| basic_edges.get(c).unwrap().clone(),
                no_end_condition
            );
            let edges: Vec<_> = edges
                .nodes
                .iter()
                .enumerate()
                .skip(1)
                .filter(|&(_, (c, _))| c.is_lowercase())
                .map(|(index, _)| {
                    // get full path to key
                    let path = edges.path(index);
                    let last = path.last().unwrap().clone();

                    // separate doors from keys
                    let (keys, intervening_keys): (Vec<_>, Vec<_>) = path
                        .into_iter()
                        .skip(1)
                        .map(|step| step.state)
                        .partition(|c| c.is_uppercase());

                    let keys = hash(keys);
                    let intervening_keys = hash(intervening_keys);

                    (
                        Key {
                            current: last.state,
                            preceding: keys,
                            intervening: intervening_keys,
                        },
                        last.cost
                    )
                })
                .collect();
            (key, edges)
        })
        .collect();

    // initial state for the main Dijkstra algorithm
    let start_state = State { location: robots, keys: [false; 26] };

    // continue Dijkstra until all keys have been collected
    let end_condition = | _: EdgeInfo<usize>, state: &State | {
        state.keys.iter().all(|&k| k)
    };

    let find_edges = | _: EdgeInfo<usize>, state: &State | -> Vec<(State, usize)> {
        let clumped: Vec<_> = state.location.iter().enumerate()
            .map(|(robot, location)| {
                key_edges.get(location).unwrap().iter()
                    .filter(|&&(key, _)| {
                        !state.keys[to_usize(key.current)] && key.preceding.iter().enumerate()
                            .filter(|&(_, k)| *k)
                            .all(|(i, _)| state.keys[i])
                    })
                    .map(|&(key, cost)| {
                        // updates robot locations
                        let mut new_location = state.location.clone();
                        new_location[robot] = key.current;

                        let mut new_keys = state.keys.clone();
                        for (i, _) in key.intervening.iter().enumerate()
                            .filter(|&(_, k)| *k)
                        {
                            new_keys[i] = true;
                        }
                        let new_state = State { location: new_location, keys: new_keys };
                        (new_state, cost)
                    })
                    .collect::<Vec<_>>()
            })
            .collect();
        clumped.into_iter().flatten().collect()
    };
    let search = dijkstra(start_state, find_edges, end_condition);

    search.steps().unwrap()
}

fn to_usize(c: char) -> usize {
    if c.is_uppercase() {
        c as usize - 'A' as usize
    } else {
        c as usize - 'a' as usize
    }
}

fn hash(i: impl IntoIterator<Item = char>) -> CharSet {
    let mut cs = [false; 26];
    for c in i {
        cs[to_usize(c)] = true;
    }
    cs
}

fn part1(tunnels: &Input) -> Output {
    solve(tunnels, vec!['@'])
}

fn part2(tunnels: &Input) -> Output {
    let robots = vec!['@', '$', '%', '^'];
    let mut quadrants = tunnels.clone();
    let original_start = quadrants.iter().position(|&c| c == '@').unwrap();
    quadrants[original_start] = '#';
    let adj_indices: Vec<usize> = quadrants
        .adjacent(original_start, true)
        .unwrap()
        .map(|adj| adj.index)
        .collect();
    for &pos in adj_indices.iter().skip(1).step_by(2) {
        quadrants[pos] = '#';
    }
    for (index, &pos) in adj_indices.iter().step_by(2).enumerate() {
        quadrants[pos] = robots[index];
    }
    solve(&quadrants, robots)
}

#[test]
fn default() {
    let input = get_input(19, 18).unwrap();
    let input = parse_input(&input);
    assert_eq!(3918, part1(&input));
    assert_eq!(2004, part2(&input));
}

// Input parsed (60μs)
// 1. 3918 (15ms)
// 2. 2004 (30ms)
// Total: 46ms